### ISSUE TITLE
C_NUMBER_RE has incorrect behavior with unary `-`

### ISSUE THREAD

**Issue by joshgoebel** [highlightjs/highlight.js#2887]
That makes sense.

I modeled my regex after the `C_NUMBER_RE` which starts with `-?`:
https://github.com/highlightjs/highlight.js/blob/3a2934d7e7af99631527bf55465fa229522735b9/src/lib/modes.js#L8

_Originally posted by @Hirse in https://github.com/highlightjs/highlight.js/pull/2808#discussion_r528422486_

**Comment by joshgoebel** [comment]
We should look at how many grammars use this rule.  It's possible worth discussing if we should change our long policy on this (though I'm still against atm).  I definitely am not sure we should get into "technically unary - is an operator in this language and a part of the number in this other languages" (recent Swift point).  I think at a highlighter level we should take a consistent position and enforce it across all the grammars.

Until I saw this I thought that position was "unary `-` is not included in numbers" because it's impossible to tell a unary `-` from a binary.   So we end up mis-highlighting things like: (highlighting the - as a number when it's an operator)

```
let x = 5-3;
```

I think the correct fix is to change the C_NUMBER rule.

This is of course less problematic in well linted code with spaces:

```
let x = 5 - -3;
```

But I don't think we can count on that.

**Comment by joshgoebel** [comment]
C_NUMBER is used in 14 grammars and our other number regexes (in our mode library) do not have this issue.  It's possible some grammars do, but thats really out of scope here without a full review... If we agree C_NUMBER is indeed in the wrong then we fix it, update any tests, and save the remaining grammars for whenever it comes up in the future.

**Comment by taufik-nurrohman** [comment]
Also make sure to exclude `-8` in `UTF-8`.

**Comment by joshgoebel** [comment]
Huh?  Not sure I'm following.

**Comment by taufik-nurrohman** [comment]
Ah okay so you want to exclude `-` from the number token globally.

**Comment by joshgoebel** [comment]
> Ah okay so you want to exclude - from the number token globally.

Yes, that has been the general rule because it's  mostly impossible to simply know if the `-` is an operator or a unary on a number (without being a full parser).

We can't look backwards yet... so all we see is "-3".. so is that:

```
[operator]- [number]3
OR
[number]-3
```

Just flagged it as an operator always (or not flagging it at all) has the appears of us making fewer 'mistakes'.

**Comment by joshgoebel** [comment]
I made the change (just fixing the mode) but it leads to +12 broken markup tests across different languages that it looks like (because of their grammar) that they could actually get away with a unary, such as say YAML:

```
value: -32
```

Where from context it's clear the `-` is associated with the number.  Also technically this change would be breaking.  Closing this for now and we'll continue to deal with numeric issues one at a time as they come up in the future.

Our overall trend has been getting away from using things like C_NUMBER_RE and instead using much more specific grammar rules per grammar.
